pandas/core/reshape/pivot.py

```diff
@@ -253,6 +252,9 @@ def _add_margins(table, data, values, rows, cols, aggfunc,
 
     row_names = result.index.names
     try:
+        for dtype in set(result.dtypes):
```

Comment:
this would break if ``dtype`` == ``category``

instead use ``select_dtypes`` to get the columns you need, then you  do
``margin_dummy[cols]  = margin_dummy[cols].astype(dtype)``